### DISCUSSION TITLE
Marketplace Updates...

### DISCUSSION THREAD

**Discussion by sbwalker** [oqtane/oqtane.framework#2981] [General]
An update on the Oqtane Marketplace....

The https://www.oqtane.net site was migrated to a dedicated App Service and SQL Database this week (previously it was part of a multi-tenant installation which included a number of other sites). All of the existing data was preserved in the migration.

A variety of  modifications were made to the Marketplace application:

- the theme from https://www.oqtane.org was modified to include a Marketplace link in the main navigation and is now being utilized on the Marketplace site for consistency
- the main landing page for the Marketplace includes a listing of products. The listing was overhauled to include the filters on the left-hand side, keyword search, and the ability to sort by a variety of different criteria. 
- The products are now displayed in grid view containing all of the previous information as well as a logo for each product
- a Submit Product button was added for developers who want to list their product in the Marketplace
- the cosmetics/styling of this page are still a work in progress
- you will notice that some existing products have display issues because they used a namespace for their product name, rather than a friendly name.

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/53f8283d-2422-4df1-8455-4537d843e277)

The Product Submission form was completely overhauled to provide a more fluid user experience with less modal popups. Product information was organized into collapsible sections:

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/2d737d78-eb70-40f1-a9ee-184c516b49e1)

Product Details now have the ability to upload a logo for a product.

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/a9b84848-208b-4ec7-838b-45f96b68586b)

License has a requirement to include a Support Url (ie. a location where a user can get support for the product).

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/b4f7449e-978f-48a8-a52f-d05e12dc6f1e)

Releases are listed and include all of the most important details:

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/1c48c23b-93df-4130-a7c8-136de6341977)

When managing a specific release there is now the ability to specify the range of Oqtane framework versions which it is Compatible with:

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/4460737a-5767-4288-8775-7b84892ed838)

Vulnerabilities can be recorded if the product has a security issue:

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/4892f914-6dc8-4e31-aaf5-f8b33b566de4)

When entering a vulnerability you can specify its Severity as well as which product Releases are affected:

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/cc08e8f7-4a0c-4f2a-8707-d3a3c6be2f06)

And Product Owner information can be included as well (including the ability to create a Stripe Connect account if you plan to offer your products commercially - more on that in a subsequent post)

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/beb4a06a-7984-4a02-b351-23037fc554c5)

**Comment by sbwalker** [comment]
The ability to support Commercial Products is coming soon...

- you will be able to choose Commercial as a License Type in the Licenses section
- you can choose to use an Integrated Support Service. The Integrated Support Service will provide a Help Desk for your products on the Marketplace. Users can enter support tickets and you can manage the tickets via a Support Dashboard. Your can also choose to use your own support service by specifying a custom Url.
- Trial Offers are supported (you can specify the Trial Period)
- you can choose to use an Integrated Payment Service. The Integrated Payment Service will leverage Stripe for e-commerce processing and vendor payments, and also provides an Integrated Licensing Service. You can also choose to use your own payment/licensing service by specifying a custom Url.
- when using the Integrated Payment Service there is an ability to specify multiple Licenses, with different terms (ie. Perpetual, Fixed Term, Monthly Subscription, Yearly Subscription), prices, etc...

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/c8decebe-e9cd-4207-b0a9-78045e64c619)

The experience for the user will begin in the Oqtane installation itself - in Module Management, Theme Management, or Language Management (this is an "app store" model). Commercial products will have a button with a Price (note that the price displayed is the minimum license price if the product has multiple licenses ):

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/e9cadafd-f780-43ab-ab24-ee8e46933690)

Clicking the Price button navigates the user to the Marketplace (ie. https://www.oqtane.net). If the product has multiple licenses they can choose their preference from the list. 

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/71d556ed-8745-4c23-b41c-9c64e0280266)

If they want to proceed they can choose Checkout button. This will navigate them to Stripe (for PCI compliance all e-commerce activities including entry of credit cards are handled by Stripe):

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/17e4896d-ea37-4050-a439-2ab57e0bfcaa)

Completing the purchase will navigate the user back to their Oqtane site. Stripe will perform an automated callback to the Marketplace once the payment is confirmed, which will make the Order as completed.

**Reply by vnetonline** [reply]
Excellent work @sbwalker the race is on to build commercial products 


Challenge accepted!!!

**Reply by sbwalker** [reply]
@vnetonline I made a LOT of progress this week... but there is still a fair amount of work remaining ie. end to end testing of each commercial scenario, licensing integration and development of the Support Dashboard (which will be based on the Help Desk by @ADefWebserver). My goal is to be complete by July 14th.

**Reply by vnetonline** [reply]
Yeah but you have made very good progress this is the first framework I have seen which has such a market place and I am excited about its future so I am investing all my spare time into it … one commercial scenario I can think of is how will the framework support SaaS applications ?? 


Anyways if you need to test something let me know I can do it for you !

**Reply by leigh-pointer** [reply]
@sbwalker i am guessing the licensing component will also be updated. This is something i have been waiting to integrate, then I will be ready to release.

**Reply by sbwalker** [reply]
@leigh-pointer yes there is a new version of the licensing component which should be ready soon

**Comment by Tony20221** [comment]
How about adding a property(s) in the listings indicating if a module is open source, free, commercial and if it's commercial the prices or at least the starting price?

Also, I was expecting to see more info on the module when I click on the module name or image. Instead, it's going to the home page of the module. The link to the home page can be on the detail page.

**Comment by sbwalker** [comment]
@Tony20221 a couple weeks ago I mentioned that I hoped to have the commercial marketplace launched by July 14... however the testing, licensing integration and development of the Support Dashboard has taken longer than expected - so the official launch has not happened yet. A more realistic estimate is early August. Once it is released you will be able to see if a module or theme is open source or commercial (including price).

**Comment by sbwalker** [comment]
Marketplace announcement: https://www.oqtane.org/blog/!/66/oqtane-marketplace-launch

**Reply by leigh-pointer** [reply]
@sbwalker absolutely brilliant

**Reply by vnetonline** [reply]
What a milestone ... excellent

**Comment by vnetonline** [comment]
How does licensing work I tried to download the Licensing component but the Install button doesn't come up after clicking the $0.00 button

**Reply by sbwalker** [reply]
@vnetonline I cannot understand your question. The Licensing component is a Nuget package - what do you mean you downloaded it? And what Install button and $0.00 button are you referring to?

**Reply by vnetonline** [reply]
when I click here download button doesn't show 

![image](https://github.com/oqtane/oqtane.framework/assets/15242923/864f900d-1961-4be5-a5f8-d98aafb65edd)

what I mean to say is I have a module that is paid that needs to check if it is licensed are there methods that I can use?

**Reply by sbwalker** [reply]
@vnetonline ok, so that is the old Module Management UI - I believe when you click the button with the $0.00 price it would take you to the Checkout page on the Marketplace, where you could select your preferred license (there are 3 different licenses for this product).

**Reply by vnetonline** [reply]
Yes I picked the freemium

**Reply by sbwalker** [reply]
@vnetonline and I am assuming the Installation Id was populated and you entered your email address and clicked Checkout - correct?

**Reply by vnetonline** [reply]
> @vnetonline and I am assuming the Installation Id was populated and you entered your email address and clicked Checkout - correct?

All worked seamlessly I am running 4.0.1 on my production site maybe that's why no download button came up

**Reply by sbwalker** [reply]
And after clicking Checkout I assume you got a confirmation message saying the Order was successful? I will need to verify if an actual order was created in the system or not. I verified that the credit card payment workflow works fine... but the freemium option uses a different workflow than commercial products (ie. they do not need to be processed through Stripe)... I will investigate further.

**Reply by vnetonline** [reply]
@sbwalker Yeah I got a confirmation it also said that my order was sucessfull

**Reply by sbwalker** [reply]
@vnetonline can you share the Installation Id and Email you used so I can investigate further (send to support@oqtane.org if you dont want to share it here)

**Comment by vnetonline** [comment]
@sbwalker I just sent you the details you requested to support@oqtane.org, Let me know if I can help in any way!

**Comment by agency8** [comment]
Hi all,
I have created a couple of commercial modules for Oqtane but there is not much info on how to integrate the licencing system into a commercial module? Can't seem to find any info on it?

Can someone help me out with this? I'm sure is easy to do but I'm a bit stumped.

Thanks,
Geoff

**Reply by leigh-pointer** [reply]
@agency8 
I will build a demo module with documentation this week and put it in a repo.

**Reply by sbwalker** [reply]
@agency8 have you looked at https://github.com/oqtane/Oqtane.LicensedModule ? It needs to be updated to .NET 8 but the general concepts remain the same. I will add some additional info to the readme in that repo.

**Reply by thabaum** [reply]
Hello @agency8 Welcome!
The Marketplace Documentation is on the table for contributions to be made in Oqtane.Docs project.

I believe @sbwalker will be doing an upcoming blog or a video spotlighting this new Oqtane Marketplace commercial licensing feature?

Community still needs time itself to catch up as you can see.   Documentation and this project is 100% community powered which is anyone including you and anyone else working with the Oqtane Framework.  

If you have specific questions relating I would bring up topics about it to ask specific questions.  I think this would be helpful to our project lead who introduced the Marketplace to the Oqtane framework and anyone looking for things to document down the road relating to the Marketplace.

I am always happy the assist when I can.  I will be testing more of this area out myself in the near future so I will be able to provide better support myself after I travel past this horizon.

I believe if you ask more specific questions here in discussions on each topic more direct answers can be provided by community members for each issue you encounter.

I hope this helps, cheers!

**Reply by thabaum** [reply]
I like @leigh-pointer idea maybe something can be updated to work with this [Oqtane/Oqtane.LicensedModule](https://github.com/oqtane/Oqtane.LicensedModule) repo?  Or one like it... so we can add it to DocFX documentation as well and follow one place to update the other?  A "Show and Tell" version here in discussions I think would help as well.

**Reply by sbwalker** [reply]
https://github.com/oqtane/Oqtane.LicensedModule  is now updated to .NET 8

**Reply by sbwalker** [reply]
https://github.com/oqtane/Oqtane.LicensedModule README.txt is now updated with documentation which describes the code in the repo

**Reply by sbwalker** [reply]
@agency8 please let me know if https://github.com/oqtane/Oqtane.LicensedModule now provides the information you were looking for... or if there is additional information you require

**Reply by agency8** [reply]
Hi All,
First, sorry for the late reaction, you all know what it's like running your own business, clients come first and time is limited.

Thank you for all of the replies and comments.

In the next week or so, when I have time, I'll be trying out the new 'LicensedModule' @sbwalker has now released (thank you for the fast reaction and creation of the new version and docs!) to see if I can get it to work with the documentation you have supplied. When I first looked at it, I couldn't find any info I could use the implement it, now there is.

@leigh-pointer If you find the time to create the demo that would be great!

Thanks,
Geoff

**Reply by sbwalker** [reply]
@agency8 more details have been added to https://github.com/oqtane/Oqtane.LicensedModule. I plan to use the new Wiki module to migrate this documentation to the Oqtane Marketplace site so that it is more accessible.
